### PR TITLE
refactor: Convert ParseObject to TypeScript

### PR DESCRIPTION
**Pull request by dplewis**
## Pull Request

- Report security issues [confidentially](https://github.com/parse-community/Parse-SDK-JS/security/policy).
- Any contribution is under this [license](https://github.com/parse-community/Parse-SDK-JS/blob/alpha/LICENSE).
- Link this pull request to an [issue](https://github.com/parse-community/Parse-SDK-JS/issues?q=is%3Aissue).

## Issue

The Type Check CI job is failing.

## Approach

* Convert ParseObject.js to TS
* Fixing lint issues
* Properly handle typing in all current TS files

I updated the Type Check CI Job
1) Ensure the `.d.ts` files generate without errors
2) Lint on those `.d.ts` files (will have errors until project is fully TS)
3) Test `.d.ts` files to find compile errors (will have errors until project is fully TS)

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -17,14 +17,18 @@ jobs:
...
       - uses: actions/checkout@v3
       - run: npm ci
-      - name: Check types
-        run: npm run test:types
+      - name: Build Types
+        run: npm run build:types
+      # - name: Lint Types
+      #   run: npm run lint:types
+      # - name: Test Types
+      #   run: npm run test:types
```
Comment (mtrezza) [inline .github/workflows/ci.yml]: Why are these commented out? We should run all type tests, even if they currently fail in the CI.

Comment (dplewis) [reply .github/workflows/ci.yml]: The have been turned back on

**Review by mtrezza — APPROVED** [review]

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.1.0-beta.1](https://github.com/parse-community/Parse-SDK-JS/releases/tag/5.1.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.1.0](https://github.com/parse-community/Parse-SDK-JS/releases/tag/5.1.0)

**Pull request merged**